Sources/XiEditor/Core/CoreNotification.swift

```diff
@@ -0,0 +1,243 @@
+// Copyright 2019 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+
+extension Array where Element == [String: Any] {
+    /// A variant of the new built-in compactMap.
+    /// This variant returns an array containing the non-nil results of calling
+    /// the given transformation with each element of this sequence.
+    /// If any transform fails and returns nil, then the entire operation fails
+    /// and returns nil.
+    func xiCompactMap<ElementOfResult>(
+        _ transform: (Element) throws -> ElementOfResult?
+    ) rethrows -> [ElementOfResult]? {
+
+        // First transform all the items
+        let cmds = try! self.map(transform)
```

Comment:
Do we need to `try!` here? This will cause the app to crash when `transform` throws, right?

Reply:
That's correct about crashing.

I'm debating. I do like the "fail fast" approach of crashing. If we used the new [`compactMap`](https://developer.apple.com/documentation/swift/sequence/2950916-compactmap) introduced in Swift 5, I think that's the behaviour we'd see. 

Finally, the only way that could throw is if the block passed into this function threw. 

So personally, I'm ok with this crashing to let us know our `transform` function has issues.

Reply:
My two cents: any place where we're handling our own data (that is where the client is deserializing something that was generated by core) I'm fine with crashing if the message is malformed.

Reply:
Totally agree! I was hoping this PR would also give us a chance to talk about general approaches to the Core <-> Client interaction.... which it is.